rasa/core/policies/policy.py

```diff
@@ -53,8 +53,9 @@ class SupportedData(Enum):
 
     @staticmethod
     def trackers_for_policy(
-        policy: Union["Policy", Type["Policy"]], trackers: List[DialogueStateTracker]
-    ) -> List[DialogueStateTracker]:
+        policy: Union["Policy", Type["Policy"]],
+        trackers: Union[List[DialogueStateTracker], List[TrackerWithCachedStates]],
```

Comment:
`DialogueStateTracker` is the parent class of `TrackerWithCachedStates` and hence covers `TrackerWithCachedStates`. No need to change the type annotation here.

Reply:
it doesn't. `train` methods use properties that exist only in `TrackerWithCachedStates`

Reply:
I know. The parent class is still enough as type annotation when it can be both. That serves as warning that we need an explicit `is_instance` check with the train function before using `TrackerWithCachedStates` methods

Reply:
ok, removed it

Reply:
actually, it's return type that is the problem

Reply:
ok, you're right. Please re-add. What a stupid problem.

Reply:
returned